datastore/sql/sqliteds.nim

```diff
@@ -29,6 +29,99 @@ proc `readOnly=`*(self: SQLiteDatastore): bool
 proc timestamp*(t = epochTime()): int64 =
   (t * 1_000_000).int64
 
+const initVersion* = 0.int64
+
+method modify*(self: SQLiteDatastore, key: Key, fn: ModifyAsync): Future[?!void] {.async.} =
```

Comment:
Shouldn't this be directly used by `put`, instead of being part of the public interface?

Reply:
Not sure if I understand this comment, you mean that the `put` would call `modify` underneath? 

If so, then it's not necessary, because `put` is an unconditional write. With `put` you write data, regardless of whatever was in the store prior to the write operation. We _could_ use `modify` for it, but we would get no benefit out of doing so, since the whole promise of `modify` is to support conditional writes (depending on what's currently in the store). It would be actually a bit slower than current `put`, because in `modify` we do at least two IO operations (`put` does exactly one IO operation).

Reply:
Yes, the problem is that we're now creating separate datastore interfaces for different applications, this is not how this was designed and it would break all the composability benefits that the current design brings . 

I think here we have two options:

- Add the modify methods to the top level interface
- Move the modify into `put` and use a flag to `force` insertion

I prefer the later, because it would keep the interface consistent with the remaining stores.

Reply:
Maybe a third option is to add an overload for `put` that takes the `modify` callback, in which case it would only put if a) no value is present or b) if the value is present it succeeds with updating it.

Reply:
>Yes, the problem is that we're now creating separate datastore interfaces for different applications, this is not how this was designed and it would break all the composability benefits that the current design brings .

How that is breaking composability? `SQLiteDatastore` continues to implement the `Datastore` interface. Additionally it implements the `ConcurrentDatastore` inteface.

>Add the modify methods to the top level interface

Ok, but then I would have to implement it on all stores, some of these implementations may be challenging and for now they would simply remain unused. I find it way more reasonable to have `modify` methods in a separate interface and possibly migrate them to main `Datastore` interface once we have them implemented for all stores (`FSDatastore`, `MountedDatastore`, `TieredDatastore` and `SQLiteDatastore`).

>Move the modify into put and use a flag to force insertion

Same issue as above, `put` is implemented by all stores, so that would imply adding implementation for all of them.

>add an overload for put that takes the modify callback

This is pretty much the same as above.

Reply:
Implemented `modify` for all stores.